keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1106 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We intend to introduce a "compatibility version" option to Kubernetes control
+plane components to make upgrades safer by increasing the granularity of steps
+available to cluster administrators. The "compatibility version" is distinct from
+the binary version of a Kubernetes component, and can be used to emulate the
+behavior (APIs, features, ...) of a prior Kubernetes release version.
+
+## Motivation
+
+The notion of more granular steps in Kubernetes upgrades is attractive because
+it is more rigorous about how we step through a Kubernetes control-plane
+upgrade, introducing potentially corrupting data (i.e. data only present in N+1
+and not in N) only in later stages of the upgrade process.
+
+For example, upgrading from Kubernetes 1.30 to 1.31 while keeping the compatibility
+version at 1.30 would enable a cluster administrator to validate that the new
+Kubernetes binary version is working as desired before exposing any feature changes
+introduced in 1.31 to cluster users, and without writing and data to storage at
+newer API versions.
+
+This extra step increases the granularity of our upgrade sequence so that 
+(1) failures are more easily diagnosed (since we have more granular steps, we
+have slightly more data sheerly from which point of the upgrade lifecycle we
+achieved failure) and (2) failures are more easily auto-reverted by upgrade
+orchestration as we are taking smaller and more incremental steps forward,
+which means there is less to “undo” on a failure condition.
+
+In beta, we intend to introduce support for greater compatibility version skew
+(specifically, N-3) so that it would be possible to skip binary versions while
+still performing a stepwise upgrade of Kubernetes control-plane. For example:
+
+- (starting point) binary-version 1.28 (compat-version 1.28)
+- upgrade binary-version to 1.31 (compat-version stays at 1.28 - this is our skip-level binary upgrade)
+- keep binary-version 1.31 while upgrading compat-version to 1.29 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.30 (stepwise upgrade of compatibility)
+- keep binary-version 1.31 while upgrading compat-version to 1.31 (stepwise upgrade of compatibility)
+
+Benefits to upgrading binary version independent of compatibility version:
+
+- During an upgrade, it is possible verify the binary version before
```

Comment:
I think this is almost all true for different binaries, too - isn't it?  Assuming you are able to detect a failure at all, you could just as easily do 1.28 binary -> 1.29 binary -> 1.30 binary -> 1.31 binary.

What IS different with a skip-version binary and compat versions is that you get a higher probability of bugs in those intermediate release already being fixed, which is valuable. 

It also seems more likely that we would have bugs, unless we run all version-specified e2es against HEAD builds with all supported compat modes on a regular basis, right?  Otherwise it would be possible to accidentally break the compat mode and not see it.  (as compared to a release branch which is tightly controlled and relativel low churn).

Would we do that on every PR?  That sounds expensive...

Reply:
> I think this is almost all true for different binaries, too - isn't it? Assuming you are able to detect a failure at all, you could just as easily do 1.28 binary -> 1.29 binary -> 1.30 binary -> 1.31 binary.

The idea  I was hoping to convey in this line item is that compatibility versions make it possible to bump the binary version while the compatibility version is fixed (e.g. binaryVersion: 1.28 -> {binaryVersion: 1.29, compatVersion: 1.28}). This allow differences in the binary (bugs fixed or introduced, performance changes, ...) to be soaked before exposing the new API/feature surface area introduced in the new version.

> What IS different with a skip-version binary and compat versions is that you get a higher probability of bugs in those intermediate release already being fixed, which is valuable.

Correct. I'll call this out in the KEP.

> It also seems more likely that we would have bugs, unless we run all version-specified e2es against HEAD builds with all supported compat modes on a regular basis, right? Otherwise it would be possible to accidentally break the compat mode and not see it. (as compared to a release branch which is tightly controlled and relativel low churn).
> Would we do that on every PR? That sounds expensive...

We're looking into how exactly we'd like to do this. Cost is a concern.  (cc @alexzielenski @BenTheElder)